src/components/znkExercise/toolbox/tools/draw/znkExerciseDrawTool.directive.js

```diff
@@ -392,51 +433,133 @@
-                        if (!newDrawMode) {
+                    function _reloadCanvas() {
+                        if (scope.d.drawMode === DRAWING_MODES.NONE) {
                             return;
                         }
+                        eventsManager.registerFbListeners(currQuestion.id);
+                    }
 
-                        _setDrawMode(newDrawMode);
-                        _updateQuestionDrawMode(newDrawMode);
-                    });
+                    function _init() {
+                        canvasContainerElement = angular.element(
+                            '<div class="draw-tool-container" ' +
+                                'ng-show="d.drawMode !== d.DRAWING_MODES.NONE" ' +
+                                'ng-class="{' +
+                                '\'no-pointer-events\': d.drawMode === d.DRAWING_MODES.VIEW,' +
+                                '\'crosshair-cursor\': d.drawMode !== d.DRAWING_MODES.NONE && d.drawMode !== d.DRAWING_MODES.VIEW' +
+                                '}">' +
+                                '<canvas></canvas>' +
+                                '</div>'
+                        );
 
-                    scope.$on('$destroy', function () {
-                        eventsManager.cleanListeners();
-                    });
+                        drawer = new Drawer();
+                        eventsManager = new EventsManager();
+                    }
+
+                    function _setContextOnHover(elementToHoverOn, canvasOfElement, canvasContextName) {
+                        
+                        var onHoverCb = function () {
+                            eventsManager.killMouseEvents();
+
+                            canvasDomElement = canvasOfElement;
+                            canvasContext = canvasDomElement.getContext("2d");
+                            serverDrawingUpdater = new ServerDrawingUpdater(currQuestion.id, canvasContextName);
+
+                            eventsManager.registerMouseEvents();
+                        };
+
+                        eventsManager.registerHoverEvent(elementToHoverOn, onHoverCb);
+
+                    }
+
+                    function _setCanvasDimensions(canvasDomElement, elementToCoverDomElement) {
+                        toolBoxCtrl.isExerciseReady().then(function () {
+                            canvasDomElement.setAttribute('height', elementToCoverDomElement.offsetHeight);
+                            canvasDomElement.setAttribute('width', elementToCoverDomElement.offsetWidth);
+                        });
+
+                    }
+
+                    function addCanvasToElement(elementToCover, question) {
+                        // we clone the element defined in _init to not mess with the upcoming append function (which doesn't work multiple times using the same element)
+                        var canvasContainerElementClone = canvasContainerElement.clone();
```

Comment:
I understood what you do the clone but the name is confusing....if a new developer will enter this code he will want to work on the canvasContainerElement and not the canvasContainerElementClone....
I suggest to change the name of the canvasContainerElement that you init to canvasContainerElementInitial and then the canvasContainerElementClone will be canvasContainerElement.
For me it is more readable